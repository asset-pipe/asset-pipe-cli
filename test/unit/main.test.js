/* eslint-disable consistent-return */
/* eslint-disable global-require */

'use strict';

let Client = require('../../');

const buildServerUri = 'http://server.com:3000';

function createRequestMock(error, response, body) {
    const { PassThrough } = require('stream');
    const mockRequest = (url, callback) => {
        if (error) return callback(error);
        callback(null, response, body);
    };
    mockRequest.post = (options, callback) => {
        const resolve = () => {
            if (error) {
                return callback(error);
            }
            callback(null, response, body);
        };
        if (options.body) {
            return resolve();
        }
        const destination = new PassThrough();
        destination.on('data', () => {});
        destination.on('end', resolve);
        return destination;
    };
    return mockRequest;
}

function createJsWriterMock() {
    const { Readable } = require('stream');
    return jest.fn(() => ({
        plugin: jest.fn(),
        transform: jest.fn(),
        bundle: () => {
            const items = [{}, {}, {}, {}];
            return new Readable({
                objectMode: true,
                read() {
                    if (!items.length) {
                        return this.push(null);
                    }
                    this.push(items.shift());
                }
            });
        }
    }));
}

beforeEach(() => {
    jest.resetModules();
});

test('new Client(options) - should throw when missing buildServerUri option', () => {
    expect(() => {
        // eslint-disable-next-line no-new
        new Client();
    }).toThrowError('Expected "server" to be a uri, got undefined');
});

test('new Client(options)', () => {
    const subject = new Client({
        buildServerUri: 'http://127.0.0.1:9999'
    });

    expect(subject.buildServerUri).toEqual('http://127.0.0.1:9999');
    expect(subject.serverId).toEqual(undefined);
    expect(subject.transforms).toHaveLength(0);
    expect(subject.plugins).toHaveLength(0);
});

test('new Client(options) 2', () => {
    const subject = new Client({
        buildServerUri: 'http://127.0.0.1:1111',
        serverId: 'some-server-id'
    });

    expect(subject.buildServerUri).toEqual('http://127.0.0.1:1111');
    expect(subject.serverId).toEqual('some-server-id');
});

test('new Client(options) 3', () => {
    const subject = new Client({ buildServerUri });

    expect(subject.buildServerUri).toEqual('http://server.com:3000');
});

test('transform(transform, options)', () => {
    const subject = new Client({ buildServerUri });
    const transform = jest.fn();
    const options = jest.fn();

    subject.transform(transform, options);

    expect(subject.transforms).toEqual([{ transform, options }]);
});

test('plugin(transform, options)', () => {
    const subject = new Client({ buildServerUri });
    const plugin = jest.fn();
    const options = jest.fn();

    subject.plugin(plugin, options);

    expect(subject.plugins).toEqual([{ plugin, options }]);
});

test('uploadFeed() - files must be an array', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.uploadFeed()).rejects.toMatchSnapshot();
});

test('uploadFeed() - files must contain at least 1 item', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.uploadFeed([])).rejects.toMatchSnapshot();
});

test('uploadFeed() - files array must only contain strings', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.uploadFeed([1, true])).rejects.toMatchSnapshot();
});

test('uploadFeed() - files array must contain .css or .js filenames', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.uploadFeed(['one', 'two'])).rejects.toMatchSnapshot();
});

test('uploadFeed(files) - request error', async () => {
    expect.assertions(1);
    jest.resetModules();
    jest.doMock('request', () => createRequestMock(new Error('Fake error!')));
    jest.doMock('@asset-pipe/js-writer', () => createJsWriterMock());
    Client = require('../../');
    const fakeFiles = ['first.js'];
    const fakeOptions = {};
    const client = new Client({ buildServerUri });

    const result = client.uploadFeed(fakeFiles, fakeOptions);

    await expect(result).rejects.toThrow('Fake error!');
});

test('createRemoteBundle(sources) - request error', async () => {
    expect.assertions(1);
    jest.resetModules();
    jest.doMock('request', () => createRequestMock(new Error('Fake error!')));
    jest.doMock('@asset-pipe/js-writer', () => createJsWriterMock());
    Client = require('../../');
    const fakeSources = ['a12das3d.json', '12da321fd.json'];
    const client = new Client({ buildServerUri });

    const result = client.createRemoteBundle(fakeSources, 'js');

    await expect(result).rejects.toThrow('Fake error!');
});

test('createRemoteBundle() - missing source argument', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.createRemoteBundle()).rejects.toMatchSnapshot();
});

test('createRemoteBundle(sources) - invalid source argument', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.createRemoteBundle('asd')).rejects.toMatchSnapshot();
});

test('createRemoteBundle(sources) - source argument missing .json ext.', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(
        client.createRemoteBundle(['a1c12a45ca1ac1ca1cc1ac1ca1'])
    ).rejects.toMatchSnapshot();
});

test('createRemoteBundle(sources) - missing type', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });
    const fakeSources = ['a12das3d.json', '12da321fd.json'];

    await expect(
        client.createRemoteBundle(fakeSources)
    ).rejects.toMatchSnapshot();
});

test('createRemoteBundle(sources) - invalid type', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });
    const fakeSources = ['a12das3d.json', '12da321fd.json'];

    await expect(
        client.createRemoteBundle(fakeSources, 'fake')
    ).rejects.toMatchSnapshot();
});

test('publishAssets(tag, entrypoints) - tag must be provided', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.publishAssets()).rejects.toMatchSnapshot();
});

test('publishAssets(tag, entrypoints) - tag must be a string', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.publishAssets(1)).rejects.toMatchSnapshot();
});

test('publishAssets(tag, entrypoints) - files must be provided', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.publishAssets('podlet')).rejects.toMatchSnapshot();
});

test('publishAssets(tag, entrypoints) - files must contain at least 1 item', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.publishAssets('podlet', [])).rejects.toMatchSnapshot();
});

test('publishAssets(tag, entrypoints) - files array must only contain strings', async () => {
    expect.assertions(2);
    const client = new Client({ buildServerUri });

    await expect(
        client.publishAssets('podlet', [1, true])
    ).rejects.toMatchSnapshot();

    await expect(client.publishAssets('podlet', 1)).rejects.toMatchSnapshot();
});

test('publishAssets(tag, entrypoints) - files array must contain .css or .js filenames', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(
        client.publishAssets('podlet', ['one', 'two'])
    ).rejects.toMatchSnapshot();
});

test('publishAssets(tag, entrypoints) - request error', async () => {
    expect.assertions(1);
    jest.resetModules();
    jest.doMock('request', () => createRequestMock(new Error('Fake error!')));
    jest.doMock('@asset-pipe/js-writer', () => createJsWriterMock());
    Client = require('../../');
    const client = new Client({ buildServerUri });

    const result = client.publishAssets('podlet', ['first.js'], {});

    await expect(result).rejects.toThrow('Fake error!');
});

test('publishInstructions(tag, type, data) - request error', async () => {
    expect.assertions(1);
    jest.resetModules();
    jest.doMock('request', () => createRequestMock(new Error('Fake error!')));
    jest.doMock('@asset-pipe/js-writer', () => createJsWriterMock());
    Client = require('../../');

    const client = new Client({ buildServerUri });
    const result = client.publishInstructions('layout', 'js', [
        'podlet1',
        'podlet2'
    ]);

    await expect(result).rejects.toThrow('Fake error!');
});

test('publishInstructions(tag, type, data) - missing arguments', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(client.publishInstructions()).rejects.toMatchSnapshot();
});

test('publishInstructions(tag, type, data) - invalid tag argument', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(
        client.publishInstructions(1, 'js', ['podlet1', 'podlet2'])
    ).rejects.toMatchSnapshot();
});

test('publishInstructions(tag, type, data) - empty data', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(
        client.publishInstructions('layout', 'js')
    ).rejects.toMatchSnapshot();
});

test('publishInstructions(tag, type, data) - data not an array', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(
        client.publishInstructions('layout', 'js', {})
    ).rejects.toMatchSnapshot();
});

test('publishInstructions(tag, type, data) - bad data in array', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(
        client.publishInstructions('layout', 'js', [1, 2])
    ).rejects.toMatchSnapshot();
});

test('publishInstructions(tag, type, data) - missing type', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(
        client.publishInstructions('layout', null, ['podlet1', 'podlet2'])
    ).rejects.toMatchSnapshot();
});

test('publishInstructions(tag, type, data) - invalid type', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });

    await expect(
        client.publishInstructions('layout', 'fake', ['podlet1', 'podlet2'])
    ).rejects.toMatchSnapshot();
});

test('bundleURL(hashes, options) - js', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });
    const url = client.bundleURL(['a', 'b']);
    expect(url).toBe(
        'http://server.com:3000/bundle/fb8e20fc2e4c3f248c60c39bd652f3c1347298bb977b8b4d5903b85055620603.js'
    );
});

test('bundleURL(hashes, options) - css', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });
    const url = client.bundleURL(['a', 'b'], { type: 'css' });
    expect(url).toBe(
        'http://server.com:3000/bundle/fb8e20fc2e4c3f248c60c39bd652f3c1347298bb977b8b4d5903b85055620603.css'
    );
});

test('bundleURL(hashes, options) - js with prefix', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });
    const url = client.bundleURL(['a', 'b'], { prefix: 'http://server' });
    expect(url).toBe(
        'http://server/fb8e20fc2e4c3f248c60c39bd652f3c1347298bb977b8b4d5903b85055620603.js'
    );
});

test('bundleURL(hashes, options) - empty array returns null', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });
    const url = client.bundleURL([]);
    expect(url).toBe(null);
});

test('bundlingComplete(hashes, options) - empty hashes array returns true', async () => {
    expect.assertions(1);
    const client = new Client({ buildServerUri });
    const isComplete = await client.bundlingComplete([]);
    expect(isComplete).toBe(true);
});

test('bundlingComplete(hashes, options) - empty hashes array returns true', async () => {
    expect.assertions(1);
    jest.resetModules();
    jest.doMock('request', () => createRequestMock(null, { statusCode: 200 }));
    Client = require('../../');
    const client = new Client({ buildServerUri });
    const isComplete = await client.bundlingComplete([]);
    expect(isComplete).toBe(true);
});

test('bundlingComplete(hashes, options) - bundle does not exist returns false', async () => {
    expect.assertions(1);
    jest.resetModules();
    jest.doMock('request', () => createRequestMock(null, { statusCode: 404 }));
    Client = require('../../');
    const client = new Client({ buildServerUri });
    const isComplete = await client.bundlingComplete(['a12sd2s1a1a1323a']);
    expect(isComplete).toBe(false);
});

test('bundlingComplete(hashes, options) - request throws', async () => {
    expect.assertions(1);
    jest.resetModules();
    jest.doMock('request', () => createRequestMock(new Error('fake error')));
    Client = require('../../');
    const client = new Client({ buildServerUri });
    await expect(client.bundlingComplete(['a12sd2s1a1a1323a'])).rejects.toEqual(
        new Error('fake error')
    );
});
