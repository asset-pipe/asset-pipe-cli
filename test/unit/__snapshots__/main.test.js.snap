// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRemoteBundle() - missing source argument 1`] = `[AssertionError [ERR_ASSERTION]: Expected argument 'sources' to be an array. Instead got undefined]`;

exports[`createRemoteBundle(sources) - invalid source argument 1`] = `[AssertionError [ERR_ASSERTION]: Expected argument 'sources' to be an array. Instead got string]`;

exports[`createRemoteBundle(sources) - invalid type 1`] = `[AssertionError [ERR_ASSERTION]: Expected argument 'type' to be one of js|css. Instead got 'fake']`;

exports[`createRemoteBundle(sources) - missing type 1`] = `[AssertionError [ERR_ASSERTION]: Expected argument 'type' to be one of js|css. Instead got 'undefined']`;

exports[`createRemoteBundle(sources) - source argument missing .json ext. 1`] = `[AssertionError [ERR_ASSERTION]: Expected ALL items in array 'sources' to end with .json. Instead got a1c12a45ca1ac1ca1cc1ac1ca1]`;

exports[`publishAssets(tag, entrypoints) - files array must contain .css or .js filenames 1`] = `
[ValidationError: Invalid 'entrypoints' argument given when attempting to publish assets. [
  "one",
  "two"
]
[31m
[1] "entrypoint filename" with value "one" fails to match the required pattern: /^[a-zA-Z0-9\\/._-]+\\.(js|css)$/[0m]
`;

exports[`publishAssets(tag, entrypoints) - files array must only contain strings 1`] = `
[ValidationError: Invalid 'entrypoints' argument given when attempting to publish assets. [
  1,
  true
]
[31m
[1] "entrypoint filename" must be a string[0m]
`;

exports[`publishAssets(tag, entrypoints) - files must be an array 1`] = `[ValidationError: Invalid 'type' argument given when attempting to determine writer. "file type" must be one of [js, css]]`;

exports[`publishAssets(tag, entrypoints) - files must contain at least 1 item 1`] = `
[ValidationError: Invalid 'entrypoints' argument given when attempting to publish assets. []
[31m
[1] "entrypoint filenames" does not contain [entrypoint filename][0m]
`;

exports[`publishAssets(tag, entrypoints) - tag must be a string 1`] = `[ValidationError: Invalid 'tag' argument given when attempting to publish assets. "tag" must be a string]`;

exports[`publishAssets(tag, entrypoints) - tag must be provided 1`] = `[ValidationError: Invalid 'tag' argument given when attempting to publish assets. "tag" is required]`;

exports[`publishInstructions(tag, type, data) - bad data in array 1`] = `
[ValidationError: Invalid 'data' argument given when attempting to publish instructions. [
  1, [31m[1][0m
  2
]
[31m
[1] "0" must be a string[0m]
`;

exports[`publishInstructions(tag, type, data) - data not an array 1`] = `
[ValidationError: Invalid 'data' argument given when attempting to publish instructions. {
  [41m"bundle instruction"[0m[31m [1]: -- missing --[0m
}
[31m
[1] "bundle instruction" must be an array[0m]
`;

exports[`publishInstructions(tag, type, data) - empty data 1`] = `[ValidationError: Invalid 'data' argument given when attempting to publish instructions. "bundle instruction" is required]`;

exports[`publishInstructions(tag, type, data) - invalid tag argument 1`] = `[ValidationError: Invalid 'tag' argument given when attempting to publish instructions. "tag" must be a string]`;

exports[`publishInstructions(tag, type, data) - invalid type 1`] = `[ValidationError: Invalid 'type' argument given when attempting to publish instructions. "file type" must be one of [js, css]]`;

exports[`publishInstructions(tag, type, data) - missing arguments 1`] = `[ValidationError: Invalid 'tag' argument given when attempting to publish instructions. "tag" is required]`;

exports[`publishInstructions(tag, type, data) - missing type 1`] = `
[ValidationError: Invalid 'type' argument given when attempting to publish instructions. {
  [41m"file type"[0m[31m [1]: -- missing --[0m
}
[31m
[1] "file type" must be one of [js, css][0m]
`;

exports[`uploadFeed() - files array must contain .css or .js filenames 1`] = `[AssertionError [ERR_ASSERTION]: Expected ALL items in array 'files' to end with .js or ALL items in array 'files' to end with .css, got one, two]`;

exports[`uploadFeed() - files array must only contain strings 1`] = `[AssertionError [ERR_ASSERTION]: Expected each item in array 'files' to be a string, got 1,true]`;

exports[`uploadFeed() - files must be an array 1`] = `[AssertionError [ERR_ASSERTION]: Expected 'files' to be an array, instead got undefined]`;

exports[`uploadFeed() - files must contain at least 1 item 1`] = `[AssertionError [ERR_ASSERTION]: Expected 'files' array to contain at least 1 item]`;
